### PR TITLE
HTTP CONNECT done non-blocking

### PR DESCRIPTION
**Pull request by bagder**

### CODE REVIEW

File: lib/urldata.h [hunk 1]
```diff
@@ -892,6 +892,21 @@ struct proxy_info {
...
+  int perline; /* count bytes per line */
+  int keepon;
+  char *line_start;
+  char *ptr; /* where to store more data */
+  curl_off_t cl; /* size of content to read and ignore */
+  enum {
+    TUNNEL_INIT,    /* init/default/no tunnel state */
+    TUNNEL_CONNECT, /* CONNECT has been sent off */
+    TUNNEL_COMPLETE /* CONNECT response received completely */
+  } tunnel_state;   /* two separate ones to allow FTP */
```
Comment (MarcelRaad) [inline lib/urldata.h]: This comment is now wrong, isn't it?

Comment (bagder) [inline lib/urldata.h]: It is indeed! Thanks. The updated logic allocates a fresh struct for each CONNECT...

File: lib/http_proxy.c [hunk 2]
```diff
@@ -343,24 +364,25 @@ static CURLcode CONNECT(struct connectdata *conn,
...
+          if(s->cl) {
             /* A Content-Length based body: simply count down the counter
                and make sure to break out of the loop when we're done! */
-            cl--;
-            if(cl <= 0) {
-              keepon = FALSE;
+            s->cl--;
+            infof(data,
+                  "one byte down '%c', %" CURL_FORMAT_CURL_OFF_T " to go!\n",
+                  *s->ptr, s->cl);
```
Comment (jay) [inline lib/http_proxy.c]: Personally I'd lose this infof but at the very least wrap it in a DEBUGF

Comment (bagder) [inline lib/http_proxy.c]: Oops, yes that's a leftover from my debugging. I'll remove that completely, thanks!

**Comment by bagder** [comment]
Just because this is a slightly larger change, I'm playing it safe and will not merge it this close to the next release. Scheduled for next instead.

**Comment by dfandrich** [comment]
This commit broke builds with --disable-proxy or --disable-http

**Comment by bagder** [comment]
Thanks, fixed now!

**Pull request closed without merging**